apollo-backend/src/main/java/io/logz/apollo/status/ServiceStatusHandler.java

```diff
@@ -0,0 +1,133 @@
+        List<Service> services = serviceDao.getAllServices();
+        Map<Service, Optional<List<Group>>> serviceGroupMapResult = new HashMap<>();
+        services.forEach(service -> {
+            Optional<Pair<Service,Optional<List<Group>>>> pair = getPairOfUndeployedServiceAndGroup(service, environmentId, timeUnit, maxUndeployedTime);
+            if(pair.isPresent()) {
+                serviceGroupMapResult.put(pair.get().getKey(), pair.get().getValue());
+            }
+        });
+        return serviceGroupMapResult;
+    }
+
+    private Optional<Pair<Service,Optional<List<Group>>>> getPairOfUndeployedServiceAndGroup(Service service, int environmentId, TimeUnit timeUnit, int maxUndeployedTime) {
+        List<Group> groups;
+        if(service.getIsPartOfGroup()) {
+            groups = getUndeployedGroupsByService(service.getId(), environmentId, timeUnit, maxUndeployedTime);
+            if(groups.size() != 0) {
+                return Optional.of(new Pair<>(service, Optional.of(groups)));
+            }
+        }
+        else {
+            if(isServiceUndeployed(service.getId(), environmentId, Optional.empty(), timeUnit, maxUndeployedTime)) {
+                return Optional.of(new Pair<>(service, Optional.empty()));
+            }
+        }
+        return Optional.empty();
+    }
+
+    private List<Group> getUndeployedGroupsByService(int serviceId, int environmentId, TimeUnit timeUnit, int maxUndeployedTime) {
+        return groupDao.getGroupsPerServiceAndEnvironment(serviceId, environmentId).stream().filter(group ->
+                isServiceUndeployed(serviceId, environmentId, Optional.of(group), timeUnit, maxUndeployedTime)).collect(Collectors.toList());
+    }
+
+    private boolean isServiceUndeployed(int serviceId, int environmentId, Optional<Group> group, TimeUnit timeUnit, int maxUndeployedTime) {
+        Optional<Date> latestDeploymentDate;
+        Optional<Date> latestUpdatedDate = getLatestDeployableVersionDateByService(serviceId);
+        if(group.isPresent()) {
+            latestDeploymentDate = getLatestDeploymentDateByServiceAndGroupAndEnvironment(serviceId, environmentId, group.get().getName());
+        }
+        else {
+            latestDeploymentDate = getLatestDeploymentDateByServiceAndEnvironment(serviceId, environmentId);
+        }
+        return isDeploymentDateExpired(latestDeploymentDate, latestUpdatedDate, timeUnit, maxUndeployedTime);
+    }
+
+    private boolean isDeploymentDateExpired(Optional<Date> latestDeploymentDate, Optional<Date> latestUpdatedDate, TimeUnit timeUnit, int maxUndeployedTime) {
+        if (latestDeploymentDate.isPresent() && latestUpdatedDate.isPresent()) {
+            return getTimeDiff(latestDeploymentDate.get(), latestUpdatedDate.get(), timeUnit) > maxUndeployedTime;
+        }
+        if (latestUpdatedDate.isPresent()) {
```

Comment:
I don't get this case.
If there is no deployable version, there is nothing to deploy to - not?